File: parapet-py/parapet/trust.py
"""Byte-range trust registry for marking untrusted content.

SDK users call ``parapet.untrusted(content, source)`` to register strings
that originate from untrusted sources (RAG, user input, web search).

On request intercept, the transport layer builds an Aho-Corasick automaton
from registered strings, searches the serialized request body, and emits
byte ranges as ``X-Guard-Trust`` header spans.
"""
from __future__ import annotations

import json
from contextvars import ContextVar
from dataclasses import dataclass

__all__ = ["TrustRegistry", "TrustSpan", "get_registry"]

# Per-context registry (thread/async-safe via contextvars)
_registry: ContextVar[TrustRegistry | None] = ContextVar(
    "parapet_trust_registry", default=None
)


@dataclass(frozen=True)
class TrustSpan:
    """A byte range in the request body marked as untrusted."""
    start: int
    end: int
    source: str = "unknown"


@dataclass
class _Entry:
    """A registered untrusted string."""
    content: str
    source: str


class TrustRegistry:
    """Registry of untrusted content strings for the current request scope.

    Thread-safe via contextvars -- each async context gets its own registry.

    Usage:
        registry = TrustRegistry()
        registry.register("some RAG content", source="rag")
        spans = registry.find_spans(serialized_body_bytes)
    """

    MAX_ENTRIES = 1000  # Configurable cap per request

    def __init__(self, max_entries: int = MAX_ENTRIES) -> None:
        self._entries: list[_Entry] = []
        self._max_entries = max_entries

    def register(self, content: str, source: str = "unknown") -> str:
        """Register a string as untrusted. Returns the string unchanged.

        Args:
            content: The untrusted content string.
            source: Provenance label (e.g., "rag", "user_input", "web_search").

        Returns:
            The content string, unchanged (for inline use).

        Raises:
            ValueError: If the registry is full (> max_entries).
        """
        if not content:
            return content

        if len(self._entries) >= self._max_entries:
            raise ValueError(
                f"TrustRegistry full: {self._max_entries} entries. "
                "Increase max_entries or reduce untrusted content registrations."
            )

        self._entries.append(_Entry(content=content, source=source))
        return content

    def find_spans(self, body: bytes) -> list[TrustSpan]:
        """Find all registered untrusted strings in a serialized request body.

        Uses a simple multi-pattern search (str.find-based for correctness in v2;
        Aho-Corasick library can be swapped in for performance if needed).

        Handles JSON string escaping: searches for the JSON-escaped form of each
        registered string, since the body is JSON-serialized.

        Args:
            body: The serialized request body bytes.

        Returns:
            List of TrustSpan for each occurrence found, sorted by start offset.
        """
        if not self._entries:
            return []

        text = body.decode("utf-8", errors="replace")
        spans: list[TrustSpan] = []

        for entry in self._entries:
            # Search for JSON-escaped form of the string since the body is JSON.
            # json.dumps adds quotes, so strip them to get the escaped interior.
            escaped = json.dumps(entry.content)[1:-1]

            # Find all occurrences of the escaped string in the body
            start = 0
            while True:
                idx = text.find(escaped, start)
                if idx == -1:
                    break
                # Convert char offset to byte offset
                byte_start = len(text[:idx].encode("utf-8"))
                byte_end = byte_start + len(escaped.encode("utf-8"))
                spans.append(TrustSpan(
                    start=byte_start, end=byte_end, source=entry.source,
                ))
                start = idx + 1

        # Sort by start offset, then by end offset (longer spans first for same start)
        spans.sort(key=lambda s: (s.start, -s.end))
        return spans

    @property
    def entry_count(self) -> int:
        """Number of registered entries."""
        return len(self._entries)

    def clear(self) -> None:
        """Remove all registered entries."""
        self._entries.clear()


def get_registry() -> TrustRegistry | None:
    """Get the trust registry for the current context, if any."""
    return _registry.get()


def _ensure_registry() -> TrustRegistry:
    """Get or create the trust registry for the current context."""
    reg = _registry.get()
    if reg is None:
        reg = TrustRegistry()
        _registry.set(reg)
    return reg


def _clear_registry() -> None:
    """Clear the trust registry for the current context."""
    reg = _registry.get()
    if reg is not None:
        reg.clear()
    _registry.set(None)

File: parapet-ts/src/trust.ts
// Copyright 2026 The Parapet Project
// SPDX-License-Identifier: Apache-2.0

import type { TrustSpan } from "./types.js";

/**
 * Per-request registry tracking which strings are untrusted data payloads.
 *
 * SDK users register strings via `untrusted(content, source)`. On request
 * intercept, the transport calls `findSpans(body)` to locate those strings
 * in the serialized JSON body, then emits byte ranges as X-Guard-Trust spans.
 */
export class TrustRegistry {
  private static readonly MAX_ENTRIES = 1000;
  private readonly entries: Array<{ content: string; source: string }> = [];
  private readonly maxEntries: number;

  constructor(maxEntries: number = TrustRegistry.MAX_ENTRIES) {
    this.maxEntries = maxEntries;
  }

  /**
   * Register a string as untrusted with a provenance label.
   * Returns the string unchanged (for inline use in message construction).
   *
   * Empty strings are ignored (no-op).
   * Throws when the registry is full.
   */
  register(content: string, source: string): string {
    if (!content) return content;

    if (this.entries.length >= this.maxEntries) {
      throw new Error(
        `TrustRegistry full: ${this.maxEntries} entries. ` +
          "Increase maxEntries or reduce untrusted content registrations.",
      );
    }

    this.entries.push({ content, source });
    return content;
  }

  /**
   * Find all registered untrusted strings in a serialized JSON request body.
   *
   * For each registered string:
   * 1. JSON-escape it (the body is JSON — the string appears escaped)
   * 2. Search for the escaped form using indexOf
   * 3. Convert character offsets to UTF-8 byte offsets
   * 4. Return sorted by start offset
   */
  findSpans(body: string): TrustSpan[] {
    if (this.entries.length === 0) return [];

    const spans: TrustSpan[] = [];

    for (const entry of this.entries) {
      // JSON.stringify wraps in quotes; strip them to get the escaped interior.
      const escaped = JSON.stringify(entry.content).slice(1, -1);

      let searchFrom = 0;
      for (;;) {
        const idx = body.indexOf(escaped, searchFrom);
        if (idx === -1) break;

        const byteStart = Buffer.byteLength(body.slice(0, idx), "utf8");
        const byteEnd =
          byteStart + Buffer.byteLength(escaped, "utf8");

        spans.push({ start: byteStart, end: byteEnd, source: entry.source });
        searchFrom = idx + 1;
      }
    }

    // Sort by start offset, then by end offset descending (longer spans first
    // for same start — matches Python SDK).
    spans.sort((a, b) => a.start - b.start || b.end - a.end);
    return spans;
  }

  /** Number of registered entries. */
  get entryCount(): number {
    return this.entries.length;
  }

  /** Whether the registry has any entries. */
  get hasEntries(): boolean {
    return this.entries.length > 0;
  }

  /** Remove all registered entries. */
  clear(): void {
    this.entries.length = 0;
  }
}

/**
 * Check whether a Content-Type header value indicates JSON content.
 *
 * Matches `application/json` or any `+json` structured syntax suffix
 * (e.g., `application/vnd.api+json`). Ignores parameters like
 * `;charset=utf-8`. Case-insensitive.
 *
 * Returns false for null/undefined/non-JSON types.
 */
export function isJsonContentType(
  contentType: string | null | undefined,
): boolean {
  if (!contentType) return false;

  // Strip parameters: "application/json; charset=utf-8" → "application/json"
  const mediaType = contentType.split(";")[0].trim().toLowerCase();

  if (mediaType === "application/json") return true;

  // Match +json structured syntax suffix (RFC 6839).
  // Must be valid type/subtype form with a non-empty prefix before +json.
  const slash = mediaType.indexOf("/");
  if (slash < 1) return false;
  const subtype = mediaType.slice(slash + 1);
  return subtype.length > "+json".length && subtype.endsWith("+json");
}
